File: core/src/main/java/io/zeebe/camel/ZeebeComponent.java
package io.zeebe.camel;

import java.util.Map;
import java.util.Objects;

import io.zeebe.camel.endpoint.EndpointConfiguration;
import io.zeebe.camel.endpoint.UniversalEventHandlerEndpoint;
import io.zeebe.camel.fn.ClientSupplier;
import io.zeebe.client.ZeebeClient;
import lombok.extern.slf4j.Slf4j;
import org.apache.camel.Endpoint;
import org.apache.camel.impl.DefaultComponent;

/**
 * The zeebe core component.
 */
@Slf4j
public class ZeebeComponent extends DefaultComponent implements ClientSupplier
{
    public static final String SCHEME = "zeebe";

    private final ZeebeClient client;

    public ZeebeComponent(final ZeebeClient client)
    {
        this.client = Objects.requireNonNull(client, "client must not be null!");
    }

    @Override
    protected Endpoint createEndpoint(final String uri, final String remaining, final Map<String, Object> parameters) throws Exception
    {
        final EndpointConfiguration configuration = EndpointConfiguration.builder()
            .uri(uri)
            .remaining(remaining)
            .parameters(parameters)
            .build();
        log.info("creating endpoint configuration={}", configuration);

        Endpoint endpoint = null;
        switch (remaining) {
        case "universalEventHandler":
            endpoint = new UniversalEventHandlerEndpoint(configuration, this);
            break;
        default:
            throw new IllegalStateException(String.format("unsupported syntax: '%s'", remaining ));
        }
        setProperties(endpoint, parameters);
        return endpoint;
    }

    @Override
    public ZeebeClient getClient()
    {
        return client;
    }
}

File: extension/core/src/main/kotlin/ZeebeComponent.kt
package io.zeebe.camel

import io.zeebe.camel.endpoint.CompleteJobEndpoint
import io.zeebe.camel.endpoint.DeployModelEndpoint
import io.zeebe.camel.endpoint.StartProcessEndpoint
import io.zeebe.camel.endpoint.SubscribeJobWorkerEndpoint
import io.zeebe.client.ZeebeClient
import io.zeebe.client.api.clients.TopicClient
import io.zeebe.client.api.events.JobEvent
import io.zeebe.client.impl.ZeebeClientImpl
import io.zeebe.client.impl.data.ZeebeObjectMapperImpl
import mu.KLogging
import org.apache.camel.Endpoint
import org.apache.camel.impl.DefaultComponent
import java.util.function.Supplier

/**
 * The zeebe core component. It is identified by camel via the {@link ZeebeComponent#SCHEME}
 * keyword. To get registered, this class is bound to the scheme in <code>/META-INF/services/org/apache/camel/component/zeebe</code>.
 */
class ZeebeComponent(private val clientSupplier: Supplier<ZeebeClient>) : DefaultComponent() {

  constructor(client: ZeebeClient) : this(Supplier { client })

  companion object : KLogging() {
    const val SCHEME = "zeebe"
  }

  private fun createEndpoint(context: ZeebeComponentContext): Endpoint = when (context.remaining) {
    CompleteJobEndpoint.COMMAND -> CompleteJobEndpoint(context)
    DeployModelEndpoint.COMMAND -> DeployModelEndpoint(context)
    SubscribeJobWorkerEndpoint.COMMAND -> SubscribeJobWorkerEndpoint(context)
    StartProcessEndpoint.COMMAND -> StartProcessEndpoint(context)
    else -> throw IllegalArgumentException("unkown: ${context.remaining}")
  }

  override fun createEndpoint(uri: String, remaining: String, parameters: MutableMap<String, Any>): Endpoint = createEndpoint(ZeebeComponentContext(uri, remaining, parameters, clientSupplier))
}

/**
 * The component context stores all endpoint parameters passed via dsl and provides access to the
 * ZeebeClient functions.
 */
data class ZeebeComponentContext(
    val uri: String,
    val remaining: String,
    val parameters: Map<String, Any>,
    val clientSupplier: Supplier<ZeebeClient>
) {
  fun topicClient(topicName : String? = null): TopicClient = with(clientSupplier.get()) {
    if (topicName == null)
      return topicClient()
    else return topicClient(topicName)
  }

  fun objectMapper(): ZeebeObjectMapperImpl {
    val client = clientSupplier.get()
    return if (client is ZeebeClientImpl)
      client.objectMapper
    else
      ZeebeObjectMapperImpl()
  }

  fun jobEvent(event: JobEvent) = objectMapper().toJson(event)!!
  fun jobEvent(json: String) = objectMapper().fromJson(json, JobEvent::class.java)!!
}
